Configurable meta-data updates to ww3_ounf

ukmo-ccbunney opened the pull request:
Fixes #272

This PR provides additional configurable metadata updates to `ww3_ounf`. A summary of the changes and new features:
  - Alternative variable type (DOUBLE, INT64) and units ("seconds from", "days from") for time variables.
  - Configurable epoch for use in time variable encoding (namelist option only).
  - Configurable "coordinate reference system" with new defaults
  - Addition of optional "forecast variables"; `forecast period` and `forecast reference time`
  - Configurable "template strings" added to facilitate different metadata for different partition numbers in partitioned output parameters.
  - Some fixes to incorrect default metadata (provided by @CarstenHansen  and closes #274)
  - Update to some "standard_names" as per ongoing work by @ukmo-ansaulter (see #280); this includes setting some standard names to be a blank string where not CF standard name is available. A "ounfmeta_legacy.inp" file is provided to maintain the existing standard_names, if required.
   - Change to "time" variable "long_name" and "conventions" attributes to reflect new encoding options.
  
To achieve the above, a significant amount of code was refactored:

 `w3ounfmeta.ftn`: One major change is the addition of a new `w3metamd.ftn` module that provides a linked list implementation for storing lists of attribute name/value pairs.

 `ww3_ounf`: The tests for SMC grid in the long IF/ELSE field block have been factored out into a subroutine; this makes the IF/ELSE tree much tidier.

There have also been some updates to the `w3timemd` module to provide time difference calculations in seconds, rather than day fractions.

ukmo-ccbunney committed:
Factored out SMC grid handling.

ukmo-ccbunney committed:
Added subroutine header and fixed indentation

ukmo-ccbunney committed:
Added flag (hard coded at moment) to output U/V vector field as
magnitude and direction. Currently only imlpemented for CUR and WND.
Do we want for other fields (such as friction velocities, etc?)

ukmo-ccbunney committed:
Added user defined partitioned parameter template strings

ukmo-ccbunney committed:
Bug fix to max partition count test

ukmo-ccbunney committed:
A couple of changes:
  - NEXT_LINE now checks for start of new section
  - User can now add partition template strings in ounfmeta.inp file

ukmo-ccbunney committed:
Added support for "coordinate reference system"

ukmo-ccbunney committed:
A few changes:
  - Better error handling when reading TEMPLATE header
  - Removed some redundant code
  - Remove all trailing whitespace

ukmo-ansaulter committed:
Standard name revisions for CF compliance

Bugfix for SNAMEP array

Bugfix2 for w3ounfmetamd

Bugfix3 for w3ounfmetamd

Bugfix4 for standard names in w3ounfmetamd

ukmo-ansaulter committed:
Added legacy standard names input file

ukmo-ccbunney committed:
Added new w3metamd module that contains definition of META_PAIR_T, META_LIST_T and associated functions. These provide a attribute name/value pair type (META_PAIR_T) and a linked list of meta pair types implemenation (META_LIST_T).
The CRS and Global meta data have been changed in w3ounfmetamd to use the new linked list rather than a flat array.

ukmo-ccbunney committed:
EXTRA user meta data for output fields is now a linked list - i.e. more
than one value can now be supplied by the user.

TODO: Need to handle the merging of partition template strings in extra
data as this likely will involve a deep copy of the linked list

ukmo-ccbunney committed:
Added "deep copy" ability to linked list; allows the "EXTRA" meta data
to be merged with partition templates without overwriting the origial
meta data entires.

ukmo-ccbunney committed:
Tidy up

ukmo-ccbunney committed:
Added default rotated pole coordinate reference system (replaces the
existing implementation in ww3_ounf)

ukmo-ccbunney committed:
Improved error handling and warnings for CRS sections - warning produced
if user overrides the "default" CRS associated with the grid.

ukmo-ccbunney committed:
Used parameter variables for <IPART> and <SPART> string tokens

ukmo-ccbunney committed:
Added "forecast period" output ("forecast reference time" is set to first
output time for now).

ukmo-ccbunney committed:
Added forecast_reference_time

ukmo-ccbunney committed:
Added support for diffent variable type and units for time variables

ukmo-ccbunney committed:
changed forecast_period variable to be int

ukmo-ccbunney committed:
Update info/error message output and clarified comments in inp/nml file

ukmo-ccbunney committed:
Fixed incorrect format statement

ukmo-ccbunney committed:
Dev comments

ukmo-ccbunney committed:
Bug fixes from @CarstenHansen.

ukmo-ccbunney committed:
Added support for user defined time epoch (for encoding of nerCDF time
variables). Also fixed bug where int64 time offset calculation may
cause integer overflow.

ukmo-ccbunney committed:
Fixed missing CRS name for standard lat/lon grid and added attributes to
specify a default "spherical earth" CRS.

ukmo-ccbunney committed:
Removed some debug output

ukmo-ccbunney committed:
Better handling of partitioned parameter strings.

ukmo-ccbunney committed:
Added ounfmeta.inp option to disable the default WW3 GLOBAL metadata

ukmo-ccbunney committed:
Updated comments and modification dates

ukmo-ccbunney committed:
Updated manual entry for ww3_ounf w.r.t. new configurable metadata
functionality.

ukmo-ccbunney commented:
Leaving this PR as a draft at the moment - there are a few other PRs in the queue that will probably cause a merge-conflict with this PR when they are merged into the NOAA repo.

ukmo-ccbunney committed:
Removed new forecast variables and time type options from ww3_ounf.inp
(will only be available via ww3_ounf.nml)

ukmo-ccbunney committed:
Removed speed/direction options for directional output (covered in
separate PR #302)

ukmo-ccbunney committed:
Finished implemented in PRINT_META function and adde some subroutine
comments.

ukmo-ccbunney committed:
Change new TP field to use S2GRID subroutine

ukmo-ccbunney committed:
Replaced CRS line that was accidentally deleted.

ukmo-ccbunney committed:
Enabled SMC type 3/4 interpoaltion (UK Met Office only fetaure - not
documented at the moment).

ukmo-ccbunney committed:
Missing directional flag on subroutine call

ukmo-ccbunney commented:
Regression test `ww3_tnc1` has been updated to test the new features in this PR.

Output from regression tests attached: [matrix_fb_ounfupd.zip](https://github.com/NOAA-EMC/WW3/files/5974147/matrix_fb_ounfupd.zip)

Summary of differences:
  - `mww3_test_03`: Usual differences
  - `ww3_tp2.18/./work_TIDE_MPI`: Known differences in ww3_prtide_current.out
  - All other differences are in netCDF gridded output files due to the updates in this PR

ukmo-ccbunney committed:
Updated regtest ww3_tnc1 to test new WW3_OUNF metadata features.

ukmo-ccbunney committed:
Added test of forecast variables and alternate time encodings for netCDF
output in regtest ww3_tnc1

ukmo-ccbunney committed:
Added -N (namelist) option to the ww3_tnc1 regtest (new featues only
available via namelist input)

ukmo-ccbunney commented:
@mickaelaccensi and @thesser1 This one is ready for review now - at your convenience.

mickaelaccensi commented on model/ftn/w3metamd.ftn:
```diff
@@ -0,0 +1,308 @@
+
+!/ ------------------------------------------------------------------- /
+      FUNCTION META_LIST_HAS_ATTR(LIST, ATTN) RESULT(FOUND)
+!     Tests whether list conatins an entry with specified attname 
```
contains

ukmo-ccbunney replied:
Corrected :)

mickaelaccensi reviewed (approved):
great work ! I approve with a minor typo to correct and also could you just explain the idea behind TIMEREF and FCVARS ? I'm not sure to understand how to exactly use them and which values they can get but I feel it could be useful for our operational forecasting system..

ukmo-ccbunney committed:
Correct typo

ukmo-ccbunney commented:
> great work ! I approve with a minor typo to correct and also could you just explain the idea behind TIMEREF and FCVARS ? I'm not sure to understand how to exactly use them and which values they can get but I feel it could be useful for our operational forecasting system..

Thanks @mickaelaccensi 
The `TIMEREF` and `FCVARS` namelist entries both relate to adding "forecast variables" to the output file.
Essentially, when `FCVARS = T`, the following variables will be added to the output file:
  - "forecast_reference_time" : The time associated with the "analysis" of the current forecast. Useful for files that are split over several days as it allows your forecast at a particular time to be referenced back to it's analysis or start time.
  - "forecast_period" : the time period elapsed (in seconds) since the associated "forecast_reference_time". Could be negative for output files that are associated with a "hindcast".

By default, the forecast variables are not output (`FCVARS = F`), to maintain the current file structure.

"forecast_reference_time" defaults to the "start time" specified in ww3_ounf.nml. However, it might be the case that the model run includes some sort of hindcast period in which case the "forecast_reference_time" would be set to the appropriate analysis time. Also, if ww3_ounf is invoked with a "start time" that is not the same as the model start time, it would be required in this case to explicitly specify the correct "forecast_reference_time" to ensure the "forecast_period" is correctly calculated.

mickaelaccensi commented:
> > great work ! I approve with a minor typo to correct and also could you just explain the idea behind TIMEREF and FCVARS ? I'm not sure to understand how to exactly use them and which values they can get but I feel it could be useful for our operational forecasting system..
> 
> Thanks @mickaelaccensi
> The `TIMEREF` and `FCVARS` namelist entries both relate to adding "forecast variables" to the output file.
> Essentially, when `FCVARS = T`, the following variables will be added to the output file:
> 
>     * "forecast_reference_time" : The time associated with the "analysis" of the current forecast. Useful for files that are split over several days as it allows your forecast at a particular time to be referenced back to it's analysis or start time.
> 
>     * "forecast_period" : the time period elapsed (in seconds) since the associated "forecast_reference_time". Could be negative for output files that are associated with a "hindcast".
> 
> 
> By default, the forecast variables are not output (`FCVARS = F`), to maintain the current file structure.
> 
> "forecast_reference_time" defaults to the "start time" specified in ww3_ounf.nml. However, it might be the case that the model run includes some sort of hindcast period in which case the "forecast_reference_time" would be set to the appropriate analysis time. Also, if ww3_ounf is invoked with a "start time" that is not the same as the model start time, it would be required in this case to explicitly specify the correct "forecast_reference_time" to ensure the "forecast_period" is correctly calculated.

thanks for this clear explanation. So it confirms my thoughts and that's typically what we need for our forecast system. Could you maybe add this explanation in the manual and a summary directly in the ww3_ounf.nml ?

JessicaMeixner-NOAA commented:
The regtests on NCEP gave similar results: 

[matrixCompFull.out.txt](https://github.com/NOAA-EMC/WW3/files/6022589/matrixCompFull.out.txt)

ukmo-ccbunney committed:
Added some extra documetation on "forecast_reference_time" and
"forecast_period" auxiliary variables in ww3_ounf.

ukmo-ccbunney commented:
> So it confirms my thoughts and that's typically what we need for our forecast system. Could you maybe add this explanation in the manual and a summary directly in the ww3_ounf.nml ?

@mickaelaccensi - I have added some extra explanation to the manual and ww3_ounf.nml as suggested.

CarstenHansen commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -1941,320 +2589,341 @@
       META(1)%ENAME  = '.ef'
       META(1)%VARNG = META(1)%VARNS
       META(1)%VMIN = -12
       META(1)%VMAX = -12.8
```
Excuse my interruption, but META(1)%VMAX = -12.8 is an error. I think it was meant to be META(1)%VMAX = 12.8. This max value seems reasonable for netCDF data type short (NCVARTYPE==2), but I don't feel sure this value is large enough for netCDF data type real (NCVARTYPE==4).

CarstenHansen replied:
Because the log10 is taken for netCDF data type short only, there must be different limits `VMIN, VMAX`. For data type real the limits of `ef` can be `META(1)%VMIN = 0; META(1)%VMAX = 1.E12`, and for data type short the variable is `log10(ef+1.E-12)` with limits `META(1)%VMIN = -12; META(1)%VMAX = 12`.

The same issue applies to parameter `P2L`.

ukmo-ccbunney replied:
Hi @CarstenHansen - you are right of course. Originally, I strived to keep the metadata the unchanged for as much of the output as possible, but in this case I managed to get the sign wrong on the VMAX. It would also be better if, as you suggest, there are different limits depending on the NCVARTYPE (due to the log10 scaling). I will make an update in this branch to fix that.

ukmo-ccbunney replied:
Fixed in commit b5a3705

ukmo-ccbunney committed:
Modified VMAX and VMIN for variables EF and P2L, as suggested by @CarstenHansen

ukmo-ccbunney commented:
Please ignore all the commits above - they are just the result of me "rebasing" on top of develop.

ukmo-ccbunney committed:
Merge branch 'develop' into fb_ounfupd

ukmo-ccbunney committed:
Increased size of ENAME and FNAMENC strings to allow longer (and more descriptive) filenames

ukmo-ccbunney committed:
Stopped rotated pole info being written out if grid is on a standard
pole, but model is compiled with RTD switch.

ukmo-ccbunney commented:
I've pushed a few extra commits to fix some issues in ww3_ounf that I found during some recent testing:
  - 6e45f0f  increases the length of the ENAME and FNAMENC strings to allow for filenames with longer field name descriptors required for CMEMS outputs,  e.g. `ww3.field_tertiary_swell_wave_directional_spread.nc`
  - f57debf adds some logic to stop extra rotated pole specific variables from being written out when the `!/RTD` switch is used but the pole location is still a standard pole.

ukmo-ccbunney committed:
Disable rotated pole flag for SMC output when SMCTYPE is 3 or 4

aliabdolali commented:
@ukmo-ccbunney Hi, please merge NOAA/develop and let me know once you solved the conflicts, then I'll run the tests on our HPC.
Thanks

ukmo-ccbunney committed:
Merge branch 'develop' into fb_ounfupd

ukmo-ccbunney commented:
@aliabdolali I've merged in the `develop` and fixed the expected merge conflicts (`w3ounfmeta.ftn` and `ww3_ounf.ftn`).
Output from rerun of regressions tests is here:

[matrix_fb_ounfupd_run2.zip](https://github.com/NOAA-EMC/WW3/files/6509529/matrix_fb_ounfupd_run2.zip)

Differences (other than in the usual non-identical tests) are all as expected and limited to the output netCDF files.

One question - I have added a default "coordinate reference system" for all gridded netCDF output files. Is everyone happy with this, or would they prefer it to not be set by default? That question aimed at @mickaelaccensi, @thesser1  and @JessicaMeixner-NOAA too...

aliabdolali commented:
> @aliabdolali I've merged in the `develop` and fixed the expected merge conflicts (`w3ounfmeta.ftn` and `ww3_ounf.ftn`).
> Output from rerun of regressions tests is here:
> 
> [matrix_fb_ounfupd_run2.zip](https://github.com/NOAA-EMC/WW3/files/6509529/matrix_fb_ounfupd_run2.zip)
> 
> Differences (other than in the usual non-identical tests) are all as expected and limited to the output netCDF files.
> 
> One question - I have added a default "coordinate reference system" for all gridded netCDF output files. Is everyone happy with this, or would they prefer it to not be set by default? That question aimed at @mickaelaccensi, @thesser1 and @JessicaMeixner-NOAA too...

Awesome, I also submitted the jobs to check regtests on our HPC. I'll have a comparison tonight or tomorrow morning. 
What would be that default "coordinate reference system"? WGS84 ellipsoid, sphere, etc?

ukmo-ccbunney commented:
Currently, the following default CRS will be added to the output file:

```
char crs ;
	crs:grid_mapping_name = "latitude_longitude" ;
	crs:longitude_of_prime_meridian = 0.f ;
	crs:semi_major_axis = 6378137.f ;
	crs:inverse_flattening = 298.2572f ;

```

See: https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/apf.html

JessicaMeixner-NOAA commented:
Can a default be overwritten? @aliabdolali  brings up a good point that some of the curvilinear polar grids have special cases but otherwise I think that sounds reasonable to me.

ukmo-ccbunney commented:
@JessicaMeixner-NOAA
> Can a default be overwritten? 

Absolutely! :)

mickaelaccensi commented:
no objection for me since it's part of CF convention

aliabdolali commented:
Hi @ukmo-ccbunney 
Here is the summary of the comp.matrix on our hpc using intel.
Could you confirm the known and expected non-b4b identical cases:

```
mww3_test_03/./work_PR2_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (7 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_tic1.4/./work_IC1IS2_1000                     (3 files differ)
ww3_tic1.4/./work_IC2IS2_IC2d                     (3 files differ)
ww3_tic1.4/./work_IC0IS2_1000                     (3 files differ)
ww3_tic1.4/./work_IC2IS2_IC2b                     (3 files differ)
ww3_tic2.2/./work_IC2IS2                     (2 files differ)
ww3_tnc1/./work                     (14 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.10/./work_MPI                     (2 files differ)
ww3_tp2.11/./work_MPI                     (2 files differ)
ww3_tp2.14/./work_OASACM5                     (2 files differ)
ww3_tp2.14/./work_OASACM6                     (2 files differ)
ww3_tp2.14/./work_OASACM4                     (2 files differ)
ww3_tp2.14/./work_OASACM3                     (2 files differ)
ww3_tp2.14/./work_OASACM                     (2 files differ)
ww3_tp2.14/./work_OASICM                     (2 files differ)
ww3_tp2.14/./work_OASOCM                     (2 files differ)
ww3_tp2.14/./work_OASACM2                     (2 files differ)
ww3_tp2.15/./work_5km                     (3 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (2 files differ)
ww3_tp2.15/./work_PR3_UQ                     (2 files differ)
ww3_tp2.15/./work_MPI_5km                     (1 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (3 files differ)
ww3_tp2.16/./work_MPI                     (3 files differ)
ww3_tp2.18/./work_TIDE                     (2 files differ)
ww3_tp2.18/./work_TIDE_MPI                     (3 files differ)
ww3_tp2.8/./work_PR3_UQ                     (3 files differ)
ww3_ts1/./work_ST6                     (2 files differ)
```

**known n0n-b4b identical:**
```
* test case: mww3_test_03; test run: ./work_PR2_UQ_MPI_d2
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_pnt.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     out_pnt.hgh2 differ (binary)
* test case: mww3_test_03; test run: ./work_PR2_UNO_MPI_d2
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_pnt.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     out_pnt.hgh2 differ (binary)
* test case: mww3_test_03; test run: ./work_PR3_UQ_MPI_d2_c
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_pnt.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     out_pnt.hgh2 differ (binary)
* test case: mww3_test_03; test run: ./work_PR3_UQ_MPI_d2
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_pnt.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     out_pnt.hgh2 differ (binary)
* test case: mww3_test_03; test run: ./work_PR3_UNO_MPI_d2_c
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     ww3_outf_hgh2/ww3.68060608.hs differ.
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     ww3_outf_hgh1/ww3.68060608.hs differ.
* test case: mww3_test_03; test run: ./work_PR1_MPI_d2
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     out_pnt.hgh2 differ (binary)
* test case: mww3_test_03; test run: ./work_PR3_UNO_MPI_d2
*********************************************************
     out_grd.low3 differ (binary)
     out_grd.low1 differ (binary)
     out_grd.hgh1 differ (binary)
     ww3_outf_hgh2/ww3.68060608.hs differ.
     out_grd.hgh2 differ (binary)
     out_grd.low2 differ (binary)
     out_pnt.low2 differ (binary)
     out_grd.hgh3 differ (binary)
     out_pnt.hgh2 differ (binary)
     ww3_outf_hgh1/ww3.68060608.hs differ.
* test case: mww3_test_07; test run: ./work_PR3_UQ
*********************************************************
     mod_def.rect1 differ (binary)
     mod_def.points differ (binary)
     mod_def.zcmpl differ (binary)
```

**Expected changes for this development:**

```
* test case: ww3_tic1.4; test run: ./work_IC1IS2_1000
*********************************************************
     ww3.201201_wn.nc differ (binary)
     ww3.201201.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tic1.4; test run: ./work_IC2IS2_IC2d
*********************************************************
     ww3.201201_wn.nc differ (binary)
     ww3.201201.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tic1.4; test run: ./work_IC0IS2_1000
*********************************************************
     ww3.201201_wn.nc differ (binary)
     ww3.201201.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tic1.4; test run: ./work_IC2IS2_IC2b
*********************************************************
     ww3.201201_wn.nc differ (binary)
     ww3.201201.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tic2.2; test run: ./work_IC2IS2
*********************************************************
     ww3.201006.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tnc1; test run: ./work
*********************************************************
     ww3.201001_tp.nc differ (binary)
     ww3_grid.out differ.
     ww3.201001_vtpk_0.nc differ (binary)
     ww3.201001_vtpk_1.nc differ (binary)
     ww3_grid.nml.log not in compare directory
     ww3.201001_vhm0_2.nc differ (binary)
     ww3.201001_vt0m1.nc differ (binary)
     ww3.201001_vtpk_3.nc not in compare directory
     ww3.201001_vtpk_2.nc differ (binary)
     ww3.201001_vhm0_1.nc differ (binary)
     ww3.201001_w10.nc differ (binary)
     ww3.201001_vtm02.nc differ (binary)
     ww3.201001_vhm0.nc differ (binary)
     ww3_ounf.nml.log not in compare directory
     ww3.201001_vmdr.nc differ (binary)
     ww3.201001_vhm0_0.nc differ (binary)
     ww3_shel.nml.log not in compare directory
     ww3.201001_vhm0_3.nc not in compare directory
     ww3_ounf.out differ.
* test case: ww3_tp2.10; test run: ./work_MPI_OMPH
*********************************************************
     ww3.196806_spec.nc differ (binary)
     ww3.68060606.hs differ.
     ww3.68060606.t01 differ.
     ww3.196806.nc differ (binary)
     out_grd.ww3 differ (binary)
     out_pnt.ww3 differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.10; test run: ./work_MPI
*********************************************************
     ww3.196806.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.11; test run: ./work_MPI
*********************************************************
     ww3.196806.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASACM5
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASACM6
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASACM4
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASACM3
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASACM
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASICM
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASOCM
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.14; test run: ./work_OASACM2
*********************************************************
     ww3.201109.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.15; test run: ./work_5km
*********************************************************
     ww3.201403.nc differ (binary)
     wind.ww3 differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.15; test run: ./work_PR3_UQ_RHO
*********************************************************
     ww3.201403.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.15; test run: ./work_PR3_UQ
*********************************************************
     ww3.201403.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.15; test run: ./work_MPI_5km
*********************************************************
     wind.ww3 differ (binary)
* test case: ww3_tp2.16; test run: ./work_MPI_OMPH
*********************************************************
     ww3.19680606.nc differ (binary)
     ww3.19680607.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.16; test run: ./work_MPI
*********************************************************
     ww3.19680606.nc differ (binary)
     ww3.19680607.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.18; test run: ./work_TIDE
*********************************************************
     ww3.201902.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.18; test run: ./work_TIDE_MPI
*********************************************************
     ww3_prnc.out differ.
     ww3.201902.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_tp2.8; test run: ./work_PR3_UQ
*********************************************************
     ww3.20080310T06Z.nc differ (binary)
     ww3.20080310T07Z.nc differ (binary)
     ww3_ounf.out differ.
* test case: ww3_ts1; test run: ./work_ST6
*********************************************************
     ww3.196806.nc differ (binary)
     ww3_ounf.out differ.

```
The summary is attached here
[DIFF_ounf.zip](https://github.com/NOAA-EMC/WW3/files/6519990/DIFF_ounf.zip)

~

ukmo-ccbunney commented:
Hi @aliabdolali ,

Yes - the expected changes in your list match mine; changes to the `*.nc` and `ww3_ounf.out` files.

However, you have two differences in your runs that I do not have:
  - **ww3_tp2.10/work_MPI_OMPH:** `out_grd.ww3` and `out_pnt.ww3` differences. Although, I believe this regtest was not b4b on your machine before?
  - **ww3_tp2.15; test run: ./work_5km:** `wind.ww3` is different. I can't explain this one - although none of the changes I have made in this PR would have affected the wind pre-processing. Would you be able to rerun this test?

Other than that, your regtests look good to me.
C

aliabdolali commented:
> Hi @aliabdolali ,
> 
> Yes - the expected changes in your list match mine; changes to the `*.nc` and `ww3_ounf.out` files.
> 
> However, you have two differences in your runs that I do not have:
> 
> * **ww3_tp2.10/work_MPI_OMPH:** `out_grd.ww3` and `out_pnt.ww3` differences. Although, I believe this regtest was not b4b on your machine before?
> * **ww3_tp2.15; test run: ./work_5km:** `wind.ww3` is different. I can't explain this one - although none of the changes I have made in this PR would have affected the wind pre-processing. Would you be able to rerun this test?
> 
> Other than that, your regtests look good to me.
> C

@ukmo-ccbunney I think the tp2.10 and tp2.15 (wind.ww3) were not b4b in the past on our hpc. I did a check for the develop.
I am going to do my final review and then we are good to go with the merge and move to the next PR.

ukmo-ccbunney committed:
Fixed incorrect valid_max for latitude variable. Fixes: #376

aliabdolali commented:
@ukmo-ccbunney 
I did check ww3_tp2.15 wind.nc. It is not related to this development, so I am going to merge #305.
For ww3_tp2.15, on our hpc using intel, the switch and switch_MPI are similar (except MPI DIST vs SHRD), and the one with MPI was identical but the SHRD one was not identical. I am going to make an issue for it.
